data_diff/diff_tables.py

```diff
@@ -176,49 +245,74 @@ def diff_tables(self, table1: TableSegment, table2: TableSegment) -> DiffResult:
+        # E.g. between id 1..10, id 5 might have been hard deleted.
+        keyspace_size = table1.end_key - table1.start_key + 1
+
+        # We only check beyond level > 0, because otherwise we might scan the
+        # entire index with COUNT(*). For large tables with billions of rows, we
+        # need to split the COUNT(*) by the `bisection_factor`.
+        if level > 0 or keyspace_size < self.bisection_threshold:
+            # In case the first segment is below the threshold
+            # This is we get the count
+            table1.compute_checksum_and_count()
+            table2.compute_checksum_and_count()
+            count1 = table1.count # These have been precomputed with the checksum
+            count2 = table2.count
+
+            # If count is below the threshold, just download and compare the columns locally
+            # This saves time, as bisection speed is limited by ping and query performance.
+            if count1 < self.bisection_threshold and count2 < self.bisection_threshold:
+                rows1 = table1.get_values()
+                rows2 = table2.get_values()
+                diff = list(diff_sets(rows1, rows2))
+                logger.info(". " * level + f"Diff found {len(diff)} different rows.")
+                yield from diff
+                return
+
+        # Find checkpoints between the two tables, e.g. [250, 500, 750] for a
+        # table with 1000 ids and a bisection factor of 4.
+        checkpoints = table1.choose_checkpoints(bisection_factor)
         assert checkpoints
-        mutual_checkpoints = table2.find_checkpoints([Value(c) for c in checkpoints])
-        mutual_checkpoints = list(set(mutual_checkpoints))  # Duplicate values are a problem!
-        logger.debug(". " * level + f"Found {len(mutual_checkpoints)} mutual checkpoints (out of {len(checkpoints)}).")
-        if not mutual_checkpoints:
-            raise Exception("Tables are too different.")
 
         # Create new instances of TableSegment between each checkpoint
-        segmented1 = table1.segment_by_checkpoints(mutual_checkpoints)
-        segmented2 = table2.segment_by_checkpoints(mutual_checkpoints)
-        if self.debug:
-            logger.debug("Performing sanity tests for chosen segments (assert sum of fragments == whole)")
-            assert count1 == sum(s.count for s in segmented1)
-            assert count2 == sum(s.count for s in segmented2)
+        # [1..249, 250..499, 500..749, 750..1000]
+        segmented1 = table1.segment_by_checkpoints(checkpoints)
+        segmented2 = table2.segment_by_checkpoints(checkpoints)
 
         # Compare each pair of corresponding segments between table1 and table2
         for i, (t1, t2) in enumerate(safezip(segmented1, segmented2)):
-            logger.info(". " * level + f"Diffing segment {i+1}/{len(segmented1)} of size {t1.count} and {t2.count}")
+            logger.info(". " * level + f"Diffing segment {i+1}/{len(segmented1)} keys={t1.start_key}..{t1.end_key-1}")
```

Comment:
We have to remove all the counts because they're too expensive, that's why we're just writing the keyspace here. I want to do that before computing the checksum in case it's slow, but we could move it after